.repo/sample-list-generator/src/libs/api-detector.ts

```diff
@@ -0,0 +1,174 @@
+import { ApiItem, ApiTypeResult, ExtensionApiMap } from '../types';
+import * as babel from '@babel/core';
+import fs from 'fs/promises';
+import { getAllFiles } from '../utils/filesystem';
+import { singularize } from '../utils/string';
+import { loadExtensionApis } from './api-loader';
+import os from 'os';
+import { ParallelController } from '../utils/parallel';
+
+const CPU_CORES = os.cpus().length;
+
+let EXTENSION_API_MAP: ExtensionApiMap = loadExtensionApis();
+
+export const getApiType = (
+  apiCategory: string,
+  apiName: string
+): ApiTypeResult => {
+  apiCategory = apiCategory.replace(/_/g, '.');
+
+  if (EXTENSION_API_MAP[apiCategory]) {
+    const apiTypes = EXTENSION_API_MAP[apiCategory];
+
+    for (let type of Object.keys(apiTypes)) {
+      if (apiTypes[type].includes(apiName)) {
+        return type as ApiTypeResult;
+      }
+    }
+  }
+  console.log('api not found', apiCategory, apiName);
+  return 'unknown';
+};
+
+const getNextLevelProperty = (
+  path: babel.NodePath<babel.types.MemberExpression>
+) => {
+  const parentNode = path.parentPath.node;
+  if (parentNode.type === 'MemberExpression') {
+    const property = parentNode.property;
+    if (property.type === 'Identifier') {
+      return property.name;
+    }
+  }
+  return '';
+};
+
+export const getApiListForSample = async (
+  sampleFolderPath: string
+): Promise<ApiItem[]> => {
+  // get all js files in the folder
+  const jsFiles = (await getAllFiles(sampleFolderPath)).filter((file) =>
+    file.endsWith('.js')
+  );
+
+  const apis: Record<string, Set<string>> = {};
+
+  const parallelController = new ParallelController(CPU_CORES);
+  const parallelHandler = jsFiles.map(async (file) => {
+    await parallelController.start();
+    const apiCalls = await extractApiCalls(await fs.readFile(file));
+    parallelController.finish();
+    Object.keys(apiCalls).forEach((apiType) => {
+      if (!apis[apiType]) {
+        apis[apiType] = new Set();
+      }
+      apiCalls[apiType].forEach((api) => {
+        apis[apiType].add(api);
+      });
+    });
+  });
+
+  await Promise.all(parallelHandler);
+
+  const result: ApiItem[] = [];
+
+  Object.keys(apis).forEach((apiType) => {
+    apis[apiType].forEach((apiName) => {
+      result.push({
+        type: singularize(apiType) as ApiTypeResult,
```

Comment:
Why do we need `singularize` here?

Reply:
Because the getApiType function returns the raw key, such as "methods". I would like to change it to "method", perhaps I should do this conversion in the getApiType function?